substrate/frame/support/src/storage/state_diff_guard.rs

```diff
@@ -0,0 +1,613 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Feature gated since it can panic.
+#![cfg(any(feature = "std", feature = "runtime-benchmarks", feature = "try-runtime", test))]
+
+//! WARNING: This code is experimental and its API might change.
+//!
+//! # Motivation
+//!
+//! In migrations and tests it is sometimes desirable to know and restrict which storage keys
+//! change. It is also helpful to get the state difference, to ensure that there are no unexpected
+//! storage changes. This module provides a guard that asserts every storage entry that has been
+//! mutated is whitelisted.
+//!
+//! # How it works
+//!
+//! When the guard is instantiated, it takes the current storage state snapshot. When the guard is
+//! dropped, it reads every storage entry and compares it with the snapshot to collect any storage
+//! entry has been changed, removed or added. It then asserts that those mutations have the
+//! one of the whitelisted prefixes.
+//!
+//! # Example
+//!
+//! Use the `StateDiffGuard` in a migration:
```

Comment:
IMO this can not be used in any migration with a reasonable state, because of the behavior of reading the *entire state*. 

This would take ages.